feat(product): add SKU search support to admin API

bqst opened the pull request:
Closes #15239

## Summary

**What** — What changes are introduced in this PR?

This PR adds the ability to search products by their variant SKU through the admin `/admin/products` endpoint using the q query parameter. Products can now be found by searching for exact or partial SKU matches across all product variants.
Note: This feature was available in [Medusa v1](https://docs.medusajs.com/v1/api/admin#products_getproducts) but was missing in v2, causing friction for users migrating from v1 or integrating with external systems.

**Why** — Why are these changes relevant or necessary?  

**Critical for WMS/ERP Integration and Internal Operations:**

- **Warehouse Management Systems (WMS) and ERP applications** typically work with SKUs as the primary product identifier for inventory management, order fulfillment, and stock tracking
- **Admin comboboxes and autocomplete** need to support SKU-based lookups for efficient product selection during order creation, inventory adjustments, and fulfillment operations
- **Customer service representatives** need to quickly locate products by SKU when handling customer inquiries or processing orders
- **Inventory managers** need to search by SKU when receiving shipments, conducting audits, or managing stock levels
- **v1 to v2 migration**: Users migrating from Medusa v1 expect this functionality to work as it did before

**How** — How have these changes been implemented?

The implementation leverages Medusa's existing free text search infrastructure:

1. **Product Model (`packages/modules/product/src/models/product.ts`)**:
   - Marked the `variants` relationship as `.searchable()` to enable search traversal into product variants
   - The free text search filter is then applied automatically by `MedusaInternalService.applyFreeTextSearchFilter` when calling `productService_.list` / `listAndCount`, so no service-layer wiring is needed (per [@NicolasGorga's review](https://github.com/medusajs/medusa/pull/13930#pullrequestreview-2895765293)).

2. **Search Behavior**:
   - Searches across product fields (title, subtitle, description) AND variant fields (SKU, title, barcode, EAN, UPC)
   - Returns the complete product (with all variants) when any variant's SKU matches
   - Supports partial matching: `?q=ABC` will find `"ABC-123"`, `"TEST-ABC-XYZ"`, etc.
   - Case-insensitive matching
   - Uses database-level full-text search for optimal performance


**Testing** — How have these changes been tested, or how can the reviewer test the feature?

Comprehensive integration tests have been added in `integration-tests/http/__tests__/product/admin/product-sku-search.spec.ts`:

✅ **All 6 tests passing:**
- ✅ Exact SKU match
- ✅ Partial SKU match
- ✅ Multi-variant products (returns full product when one variant matches)
- ✅ Combined search (works for both product title AND variant SKU)
- ✅ Non-existent SKU (correctly returns empty results)
- ✅ Case-insensitive search

**To test manually:**

```bash
# Search by exact SKU
GET /admin/products?q=SKU-123

# Search by partial SKU
GET /admin/products?q=SKU

# Combined with other filters
GET /admin/products?q=SKU-123&status=published&limit=10
```

---

## Checklist

Please ensure the following before requesting a review:

- [x] I have added a **changeset** for this PR
    - Every non-breaking change should be marked as a **patch**
    - To add a changeset, run `yarn changeset` and follow the prompts
- [x] The changes are covered by relevant **tests**
- [x] I have verified the code works as intended locally
- [x] I have linked the related issue(s) if applicable

---

## Additional Context

- No breaking changes - this is a backward-compatible enhancement
- No new dependencies required
- Performance impact is minimal as it uses existing indexed fields
- Works seamlessly with existing pagination, filtering, and sorting parameters
- Restores feature parity with Medusa v1


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Changes product listing query behavior by introducing free-text search traversal across variants, which may affect result sets and query performance on large catalogs; covered by integration tests.
> 
> **Overview**
> Adds support for free-text searching products by variant SKU (and other searchable variant fields) via the admin `GET /admin/products?q=...` endpoint by wiring `q` into the module’s free-text search filter.
> 
> This marks the `Product.variants` relation as `.searchable()` to allow search traversal into variants, and updates `listProducts`/`listAndCountProducts` to translate `filters.q` into a `FreeTextSearchFilterKeyPrefix + Product` filter (removing `q` from the remaining filters). Integration tests are added to cover exact/partial SKU matching, multi-variant results, combined title+SKU search, non-matches, and case-insensitivity.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e024914dd9b05fba7c08ebc82f85902f361b6e09. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

bqst committed:
feat(product): add SKU search support to admin API

bqst committed:
🔧 Add changeset

willbouch commented on .changeset/forty-moles-give.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@medusajs/product": minor
```
Is this really minor? If there are no breaking change (and I dont see any?), it should be a patch

bqst replied:
You’re right, it’s a non-breaking enhancement.
I’ve updated the changeset accordingly.

bqst committed:
🔧 Update changeset

bqst committed:
🔧 Update changeset

NicolasGorga reviewed (changes requested):
Hey @bqst thank you for the contribution. The free text filtering is applied automatically at a lower level by `MedusaInternalService` for `list` and `listAndCount` service factory methods. See [here](https://github.com/medusajs/medusa/blob/62e0760a2ad032c9f7ef219968d667193bf3c4a1/packages/core/utils/src/modules-sdk/medusa-internal-service.ts#L243)

Given that the `ProductVariant` already defines its sku related fields with the `searchable` modifier, the only thing needed for them to be searchable from product searches is to add the modifier to the `variants` relation, as you did. Could you please remove the update you made to the product module service, then?

NicolasGorga commented on .changeset/forty-moles-give.md:
```diff
@@ -0,0 +1,6 @@
+---
+"@medusajs/product": patch
+"integration-tests-http": patch
```
@bqst We don't include tests package in the changeset, if you coupld remove that'd be super

bqst replied:
Done, `integration-tests-http` removed from the changeset in ca7a1b6f77.

NicolasGorga committed:
Merge branch 'develop' into feat/product-sku-search

bqst committed:
refactor(product): rely on MedusaInternalService for free text search

Remove redundant FreeTextSearchFilterKeyPrefix wiring in
listProducts/listAndCountProducts: applyFreeTextSearchFilter is already
applied at the lower MedusaInternalService level, so marking
Product.variants as .searchable() is sufficient to traverse into the
already-searchable variant fields (sku, title, barcode, ean, upc).

Also drop integration-tests-http from the changeset.

bqst committed:
Merge remote-tracking branch 'upstream/develop' into feat/product-sku-search

NicolasGorga reviewed (approved):
LGTM, thanks for the contribution!
